src/internet_identity/src/state/temp_keys.rs

```diff
@@ -25,56 +39,98 @@ pub struct TempKeys {
     /// bounded by the registration rate limit.
-    temp_keys: HashMap<DeviceKey, TempKey>,
+    device_bound_keys: HashMap<DeviceKey, DeviceBoundTempKey>,
 
     /// Deque to efficiently prune expired temp keys
     expirations: VecDeque<TempKeyExpiration>,
 }
 
 impl TempKeys {
-    pub fn add_temp_key(
+    pub fn add_onboarding_temp_key(&mut self, temp_key: Principal) -> Result<(), ()> {
+        self.prune_expired_keys();
+
+        // tie the max number of onboarding keys to the registration rate limit
+        if self.onboarding_keys.len() as u64
+            > state::persistent_state(|ps| {
+                ps.registration_rate_limit
+                    .as_ref()
+                    .map(|rate_limit| rate_limit.max_tokens)
+                    .unwrap_or(20_000)
+            })
+        {
+            return Err(());
+        }
+
+        let expiration = time() + TEMP_KEY_EXPIRATION_NS;
+        self.expirations.push_back(TempKeyExpiration::Onboarding {
+            principal: temp_key,
+            expiration,
+        });
+        self.onboarding_keys.insert(temp_key, expiration);
+        Ok(())
+    }
+
+    pub fn add_device_bound_temp_key(
         &mut self,
         device_key: &DeviceKey,
         anchor: AnchorNumber,
         temp_key: Principal,
     ) {
-        let tmp_key = TempKey {
+        // Remove the temp key from onboarding keys if it exists
+        self.onboarding_keys.remove(&temp_key);
+
+        let tmp_key = DeviceBoundTempKey {
             principal: temp_key,
             anchor,
             expiration: time() + TEMP_KEY_EXPIRATION_NS,
         };
 
-        self.expirations.push_back(TempKeyExpiration {
+        self.expirations.push_back(TempKeyExpiration::DeviceBound {
             device_key: device_key.clone(),
             expiration: tmp_key.expiration,
         });
-        self.temp_keys.insert(device_key.clone(), tmp_key);
+        self.device_bound_keys.insert(device_key.clone(), tmp_key);
     }
 
     /// Removes the temporary key for the given device if it exists and is linked to the provided anchor.
     pub fn remove_temp_key(&mut self, anchor: AnchorNumber, device_key: &DeviceKey) {
         // we can skip the removal from expirations because there it will be removed
         // during amortized clean-up operations
-        if let Some(temp_key) = self.temp_keys.get(device_key) {
+        if let Some(temp_key) = self.device_bound_keys.get(device_key) {
             if temp_key.anchor == anchor {
                 // Only remove temp key if the anchor matches
-                self.temp_keys.remove(device_key);
+                self.device_bound_keys.remove(device_key);
             }
         }
     }
 
+    /// Checks that the temporary key is a valid onboarding key (i.e. has solved the captcha).
+    ///
+    /// Requires a mutable reference because it does amortized clean-up of expired temp keys.
```

Comment:
Would it make sense to lift the pruning out then? The function would be less surprising and you wouldn't need mutability